{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "96ba40a2-f1cd-4a8f-9801-877340996d17",
   "metadata": {},
   "source": [
    "# **Decoupler-GPU**\n",
    "**Accelerated Transcriptional Regulatory Analysis**  \n",
    "\n",
    "**Author:** [Severin Dicks](https://github.com/Intron7)\n",
    "**Copyright** [scverse](https://scverse.org)\n",
    "\n",
    "In this notebook, we explore the updated Decoupler functions for inferring transcriptional regulatory activity, leveraging the **GPU-accelerated** implementation from **rapids-singlecell**.\n",
    "\n",
    "By running these analyses on GPUs, we achieve faster computation and scalability for large single-cell datasets.\n",
    "We will apply **Decoupler** methods to estimate **transcription factor (TF) activity** and **pathway activity**, utilizing curated regulatory networks such as **CollecTRI**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79abebac-6067-43de-b133-e1e169d93c2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rapids_singlecell as rsc\n",
    "import scanpy as sc\n",
    "import rmm\n",
    "import cupy as cp\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import decoupler as dc\n",
    "from rmm.allocators.cupy import rmm_cupy_allocator\n",
    "rmm.reinitialize(\n",
    "    managed_memory=False, # Allows oversubscription\n",
    "    pool_allocator=True, # default is False\n",
    ")\n",
    "cp.cuda.set_allocator(rmm_cupy_allocator)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "365a072f",
   "metadata": {},
   "source": [
    "ℹ️ Note: The dataset used in this notebook is generated in `01_demo_gpu.ipynb`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ed3ba4-c931-4ed8-8773-43b6d469b30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"h5/dli_decoupler.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b659c2f8-b000-42e3-ad85-0d274d5c40c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.raw.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7074267b-5245-4819-9955-0ea40a289b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsc.get.anndata_to_GPU(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ed6ea0-1f5f-4a76-8417-71888bde94c0",
   "metadata": {},
   "source": [
    "### CollecTRI network\n",
    "CollecTRI is a comprehensive resource containing a curated collection of TFs and their transcriptional targets compiled from 12 different resources. This collection provides an increased coverage of transcription factors and a superior performance in identifying perturbed TFs compared to other literature based GRNs such as DoRothEA. Similar to DoRothEA, interactions are weighted by their mode of regulation (activation or inhibition)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6dd643c-a203-4f7f-933f-ae92331f9dba",
   "metadata": {},
   "outputs": [],
   "source": [
    "collectri = dc.op.collectri(organism=\"human\",license=\"commercial\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c45196c-5430-4d63-b432-f5eb47690393",
   "metadata": {},
   "source": [
    "### Transcription Factor Activity Inference with ULM  \n",
    "\n",
    "To infer transcription factor (TF) activity, we use the **Univariate Linear Model (ULM)** from **Decoupler**.  \n",
    "ULM estimates the regulatory influence of TFs on gene expression by fitting a linear model for each TF-gene interaction.  \n",
    "\n",
    "We apply **ULM** using the **Dorothea** regulatory network retrieved earlier. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76156284-d218-457a-84e6-b8fcd11ee5eb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.dcg.ulm(data=adata, net=collectri, verbose = True, raw=False, bsize=10000, tmin=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e97337b-73f3-449b-8f35-306c0626c557",
   "metadata": {},
   "outputs": [],
   "source": [
    "score = dc.pp.get_obsm(adata=adata, key=\"score_ulm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1456532e-0415-4a61-81bd-2623ffc2780e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(score,basis=\"X_umap_harmony\", color=[\"cell_type\", \"GATA3\"], cmap='RdBu_r',ncols=1)\n",
    "sc.pl.violin(score, keys=['GATA3'], groupby='cell_type', rotation=90, size = 0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7c436e7-bcee-419d-bbfe-0af008732e4c",
   "metadata": {},
   "source": [
    "### Transcription Factor Activity Inference with MLM  \n",
    "\n",
    "Next, we infer transcription factor (TF) activity using the **Multivariate Linear Model (MLM)** from **Decoupler**.  \n",
    "Unlike **ULM**, which considers each TF independently, **MLM** accounts for multiple TFs simultaneously,  \n",
    "providing a more comprehensive estimation of regulatory influences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "401648ad-79da-4320-995d-d76ef5e7f3f2",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.dcg.mlm(data=adata, net=collectri, verbose = True, raw=False, bsize=10000, tmin=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd41485-f31a-4fd1-889a-169d6e4ad0d8",
   "metadata": {},
   "source": [
    "### Transcription Factor Activity Inference with AUCell  \n",
    "\n",
    "Next, we infer transcription factor (TF) activity using **AUCell** from **Decoupler**.  \n",
    "AUCell estimates TF activity based on the **enrichment of TF target genes within the top-expressed genes** in each cell.  \n",
    "Unlike regression-based methods (**ULM**, **MLM**), AUCell is a **rank-based approach** that does not assume linear relationships. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83a049b5-a0f1-4bd9-adad-49df94258e54",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.dcg.aucell(data=adata, net=collectri, verbose = True, raw=False, bsize=5000, tmin=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cabf07f1-99aa-4a7b-87d4-7af5ff8100e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "score = dc.pp.get_obsm(adata=adata, key=\"score_aucell\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82660514-d0e1-4d49-956c-382d9a588212",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(score,basis=\"X_umap_harmony\", color=[\"cell_type\", \"GATA3\"], cmap='RdBu_r',ncols=1)\n",
    "sc.pl.violin(score, keys=['GATA3'], groupby='cell_type', rotation=90, size = 0)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
